src/Microsoft.Performance.Toolkit.Engine/DataSourceSet.cs

```diff
@@ -0,0 +1,537 @@
+        /// <returns>
+        ///    A reference to this instance after the operation has completed.
+        /// </returns>
+        /// <exception cref="ObjectDisposedException">
+        ///     This instance is disposed.
+        /// </exception>
+        public ReadOnlyDataSourceSet AsReadOnly()
+        {
+            this.ThrowIfDisposed();
+
+            //
+            // Create a deep copy of the given data source set. This
+            // is used to make sure that the readonly set cannot be
+            // changed by manipulating the original data source set reference.
+            // Because a DataSourceSet may own the plugin set,
+            // copies created by this method will *never* own the plugin
+            // set from the original in order to avoid ambiguity in
+            // ownership of the plugin set. This means that the result
+            // of this method is not an exact deep copy, but for the 
+            // purposes of the engine, it works. If we decide we want to
+            // expose copy functionality to the public, then we need to
+            // figure out how to deal with ownership of the underlying
+            // plugin set in regards to copies so that we can communicate
+            // them appropriately.
+            //
+
+            var dataSourcesToProcess = new Dictionary<ProcessingSourceReference, List<List<IDataSource>>>();
+            var freeDataSources = new List<IDataSource>();
+            var processingSourceReferencesList = new List<ProcessingSourceReference>();
+
+            foreach (var kvp in this.dataSourcesToProcess)
+            {
+                dataSourcesToProcess[kvp.Key] = kvp.Value
+                    .Select(x => new List<IDataSource>(x))
+                    .ToList();
+            }
+
+            freeDataSources.AddRange(this.FreeDataSourcesToProcess);
+            processingSourceReferencesList.AddRange(this.processingSourceReferencesList);
+
+            return new ReadOnlyDataSourceSet(
+                dataSourcesToProcess,
+                freeDataSources,
+                this.plugins);
+        }
+
+        private bool TypeIs(Type first, Type second)
+        {
+            Debug.Assert(!this.isDisposed);
+            Debug.Assert(first != null);
+            Debug.Assert(second != null);
+
+            if (this.Plugins.ArePluginsIsolated)
+            {
+                return first.GUID == second.GUID &&
+                       first.AssemblyQualifiedName == second.AssemblyQualifiedName;
+            }
+            else
+            {
+                return first.Is(second);
+            }
+        }
+
+        private void AddDataSourcesCore(
+            IEnumerable<IDataSource> dataSources,
+            Type processingSourceType,
+            List<ProcessingSourceReference> processingSourceReferences,
+            Dictionary<ProcessingSourceReference, List<List<IDataSource>>> dataSourcesToProcess,
+            Func<Type, Type, bool> typeIs)
```

Comment:
>typeIs [](http://example.com/codeflow?start=35&length=6)

why is 'typeIs' being passed in as a function?

Reply:
was this method originally static or something? I think several of these parameters could be removed and accessed directly from 'this'

Reply:
Weird that not all comments from the other PR tool seem to be added here. so... I'll copy/paste:

was this method originally static or something? I think several of these parameters could be removed and accessed directly from 'this'

Reply:
Yes it was static. I will look at using this.